File: platform/src/main/java/com/auraboot/framework/inbox/service/InboxService.java
package com.auraboot.framework.inbox.service;

import com.auraboot.framework.inbox.model.InboxItem;
import com.baomidou.mybatisplus.core.metadata.IPage;

import java.util.Map;

/**
 * Unified Inbox service — manages materialized inbox items
 * aggregated from BPM approvals, IM mentions, task assignments, etc.
 *
 * @since 6.3.0
 */
public interface InboxService {

    /**
     * Create a new inbox item. Supports dedup via clientItemId.
     * Returns existing item if clientItemId already exists.
     */
    InboxItem createItem(InboxItem item);

    /**
     * Get an inbox item by ID (with tenant/user ownership check).
     */
    InboxItem getItem(Long id, Long userId, Long tenantId);

    /**
     * List inbox items for a user with pagination and optional filters.
     *
     * @param userId   target user
     * @param tenantId tenant
     * @param itemType optional filter (APPROVAL, TASK, MENTION, etc.)
     * @param status   optional filter (PENDING, ACTED, DISMISSED)
     * @param pageNum  page number (1-based)
     * @param pageSize page size
     */
    IPage<InboxItem> listByUser(Long userId, Long tenantId, String itemType,
                                 String status, int pageNum, int pageSize);

    /**
     * List inbox items for a user with pagination and multi-type filter.
     * Used for segment-based queries (e.g. Pending = APPROVAL + TASK_DUE + ASSIGNMENT).
     *
     * @param userId    target user
     * @param tenantId  tenant
     * @param itemTypes optional filter — list of item types (null = all)
     * @param status    optional filter (PENDING, ACTED, DISMISSED)
     * @param pageNum   page number (1-based)
     * @param pageSize  page size
     */
    IPage<InboxItem> listByUser(Long userId, Long tenantId, java.util.List<String> itemTypes,
                                 String status, int pageNum, int pageSize);

    /**
     * Get unread counts grouped by item type.
     */
    Map<String, Integer> getUnreadSummary(Long userId, Long tenantId);

    /**
     * Total unread count.
     */
    int getUnreadCount(Long userId, Long tenantId);

    /**
     * Mark a single item as read.
     */
    void markRead(Long id, Long userId, Long tenantId);

    /**
     * Mark all pending items as read.
     */
    int markAllRead(Long userId, Long tenantId);

    /**
     * Mark an item as acted (approve, reject, mark_done, etc.).
     */
    void markActed(Long id, Long userId, Long tenantId, String action);

    /**
     * Dismiss an item (user chose to ignore it).
     */
    void dismiss(Long id, Long userId, Long tenantId);

    /**
     * Batch mark items as read.
     */
    int batchMarkRead(java.util.List<Long> ids, Long userId, Long tenantId);

    /**
     * Batch mark items as acted.
     */
    int batchMarkActed(java.util.List<Long> ids, Long userId, Long tenantId, String action);

    /**
     * Batch dismiss items.
     */
    int batchDismiss(java.util.List<Long> ids, Long userId, Long tenantId);

    /**
     * Close inbox items whose clientItemId starts with the given prefix.
     * Used to close all items for a completed/canceled task (covers old and new clientItemId formats).
     */
    void closeByClientItemIdPrefix(String prefix);

    /**
     * Close a specific inbox item by exact clientItemId.
     * Used for targeted close (e.g. transfer removes specific user's item).
     */
    void closeByClientItemId(String clientItemId);

    /**
     * Close inbox items matching prefix but excluding a specific clientItemId.
     * Used for task_claimed: close other candidates' items but keep claimer's.
     *
     * @param prefix             clientItemId prefix (e.g. "bpm_task_{taskId}_")
     * @param excludeClientItemId the clientItemId to keep (claimer's item)
     * @param reason             action_taken value (e.g. "claimed_by_other")
     */
    void closeByClientItemIdPrefixExcluding(String prefix, String excludeClientItemId, String reason);

    /**
     * Close a specific inbox item by clientItemId with a reason.
     * Used for task_revoked: close removed assignee's item with explanation.
     */
    void closeByClientItemIdWithReason(String clientItemId, String reason);
}

File: web-admin/app/shared/services/inboxService.ts
/**
 * Unified Inbox Service — API client for /api/inbox endpoints.
 *
 * Data source: ab_inbox_item table via InboxController.
 * Used by InboxBadge, InboxDropdown, and UnifiedInboxPage.
 *
 * @since 6.4.0
 */

import { fetchResult } from '~/shared/services/http-client';
import { ResultHelper } from '~/utils/type';

export interface InboxItem {
  id: number;
  tenantId: number;
  userId: number;
  itemType: string; // 'approval' | 'task' | 'mention' | 'ai_suggestion' | 'alert' | 'assignment'
  title: string;
  subtitle?: string;
  priority: string; // 'low' | 'normal' | 'high' | 'urgent'
  status: string; // 'pending' | 'acted' | 'dismissed' | 'expired' | 'closed'
  sourceType?: string; // 'bpm' | 'im' | 'command' | 'ai' | 'notification'
  sourceId?: string;
  modelCode?: string;
  recordId?: number;
  cardPayload?: string; // JSON string
  actionTaken?: string;
  actedAt?: string;
  deepLink?: string;
  isRead: boolean;
  readAt?: string;
  createdAt: string;
  expiresAt?: string;
  clientItemId?: string;
}

export interface InboxPage {
  records: InboxItem[];
  total: number;
  current: number;
  size: number;
  pages: number;
}

export interface UnreadSummary {
  [key: string]: number;
}

const BASE = '/api/inbox';

/**
 * List inbox items with optional filters.
 */
export async function listInboxItems(params: {
  itemType?: string;
  status?: string;
  pageNum?: number;
  pageSize?: number;
}): Promise<InboxPage> {
  const result = await fetchResult<InboxPage>(BASE, {
    method: 'get',
    params,
  });
  if (ResultHelper.isSuccess(result) && result.data) {
    return result.data;
  }
  return { records: [], total: 0, current: 1, size: 20, pages: 0 };
}

/**
 * Get unread counts grouped by item type.
 */
export async function getUnreadSummary(): Promise<UnreadSummary> {
  const result = await fetchResult<UnreadSummary>(`${BASE}/unread-summary`, {
    method: 'get',
  });
  if (ResultHelper.isSuccess(result) && result.data) {
    return result.data;
  }
  return {};
}

/**
 * Get total unread count (for badge).
 */
export async function getUnreadCount(): Promise<number> {
  const result = await fetchResult<number>(`${BASE}/unread-count`, {
    method: 'get',
  });
  if (ResultHelper.isSuccess(result) && result.data != null) {
    return typeof result.data === 'number' ? result.data : 0;
  }
  return 0;
}

/**
 * Get single inbox item detail.
 */
export async function getInboxItem(id: number): Promise<InboxItem | null> {
  const result = await fetchResult<InboxItem>(`${BASE}/${id}`, {
    method: 'get',
  });
  if (ResultHelper.isSuccess(result) && result.data) {
    return result.data;
  }
  return null;
}

/**
 * Get full approval detail for a BPM task inbox item.
 */
export async function getApprovalDetail(id: number): Promise<any> {
  const result = await fetchResult<any>(`${BASE}/${id}/approval-detail`, {
    method: 'get',
  });
  return ResultHelper.isSuccess(result) ? result.data : null;
}

/**
 * Mark a single item as read.
 */
export async function markRead(id: number): Promise<void> {
  await fetchResult(`${BASE}/${id}/read`, { method: 'put' });
}

/**
 * Mark all items as read.
 */
export async function markAllRead(): Promise<void> {
  await fetchResult(`${BASE}/read-all`, { method: 'put' });
}

/**
 * Mark item as acted with an action.
 */
export async function markActed(id: number, action: string): Promise<void> {
  await fetchResult(`${BASE}/${id}/act`, { method: 'put', params: { action } });
}

/**
 * Dismiss an item.
 */
export async function dismissItem(id: number): Promise<void> {
  await fetchResult(`${BASE}/${id}/dismiss`, { method: 'put' });
}

/**
 * Submit approval action on an inbox item.
 */
export async function submitApprovalAction(id: number, action: string): Promise<void> {
  await fetchResult(`${BASE}/${id}/approval-action`, {
    method: 'post',
    params: { action },
  });
}

/**
 * Batch approve items.
 */
export async function batchApprove(ids: number[]): Promise<void> {
  await fetchResult(`${BASE}/batch/approve`, { method: 'post', params: { ids } });
}

/**
 * Batch reject items.
 */
export async function batchReject(ids: number[]): Promise<void> {
  await fetchResult(`${BASE}/batch/reject`, { method: 'post', params: { ids } });
}

/**
 * Batch mark items as read.
 */
export async function batchMarkRead(ids: number[]): Promise<void> {
  await fetchResult(`${BASE}/batch/read`, { method: 'put', params: { ids } });
}
